src/typingsInstaller/nodeTypingsInstaller.ts

```diff
@@ -105,28 +106,45 @@ namespace ts.server.typingsInstaller {
             ({ execSync: this.nodeExecSync } = require("child_process"));
 
             this.ensurePackageDirectoryExists(globalTypingsCacheLocation);
+            const packageName = this.installTypesRegistry(globalTypingsCacheLocation);
+            this.typesRegistry = loadTypesRegistryFile(getTypesRegistryFileLocation(globalTypingsCacheLocation, packageName), this.installTypingHost, this.log);
+        }
 
-            try {
-                if (this.log.isEnabled()) {
-                    this.log.writeLine(`Updating ${typesRegistryPackageName} npm package...`);
-                }
-                this.execSyncAndLog(`${this.npmPath} install --ignore-scripts ${typesRegistryPackageName}@${this.latestDistTag}`, { cwd: globalTypingsCacheLocation });
-                if (this.log.isEnabled()) {
-                    this.log.writeLine(`Updated ${typesRegistryPackageName} npm package`);
+        private installTypesRegistry(globalTypingsCacheLocation: string) {
+            let result: typeof typesRegistryPackageName | typeof definitelyTypedTypesRegistryPackageName | "UPDATE FAILED" =
+                this.installTypesRegistryFromPackageName(definitelyTypedTypesRegistryPackageName, globalTypingsCacheLocation);
+            if (result === "UPDATE FAILED") {
+                result = this.installTypesRegistryFromPackageName(typesRegistryPackageName, globalTypingsCacheLocation);
+                if (result === "UPDATE FAILED") {
+                    if (this.log.isEnabled()) {
+                        this.log.writeLine(`Error updating ${typesRegistryPackageName} package`);
+                    }
+                    // store error info to report it later when it is known that server is already listening to events from typings installer
+                    this.delayedInitializationError = {
+                        kind: "event::initializationFailed",
+                        message: result
+                    };
+                    return "types-registry";
                 }
             }
-            catch (e) {
+            return result;
+        }
+
+        private installTypesRegistryFromPackageName(packageName: typeof typesRegistryPackageName | typeof definitelyTypedTypesRegistryPackageName, globalTypingsCacheLocation: string) {
+            if (this.log.isEnabled()) {
+                this.log.writeLine(`Updating ${packageName} npm package...`);
+            }
+            const registry = packageName === typesRegistryPackageName ? "" : "--registry=https://npm.pkg.github.com";
+            const failed = this.execSyncAndLog(`${this.npmPath} install ${registry} --ignore-scripts ${packageName}@${this.latestDistTag}`, { cwd: globalTypingsCacheLocation });
+            if (failed) {
+                return "UPDATE FAILED";

```

Comment:
Well, here's my thought process, such as it was:

1. I want the Typescript equivalent of `Maybe` (aka `Result`), not `Either`, because I don't have anything useful to return in the error case.
2. That's `x | undefined` usually.
3. But my return type is a string literal union, so why not just return another string? It's going to be checked precisely anyway. And it'll be slightly more efficient and explicit.
4. Plus I get to come up with a hilarious failure string: BAMF? TRY AGAIN LATER? DON'T COPY THAT FLOPPY?

My baseline state is just dislike of exceptions so I didn't think of them.
In any case, the fallback-with-two-tries structure seems like it would be even clunkier with exceptions, although I could be wrong.